src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -165,90 +190,84 @@ library GPv2Encoding {
-        // However, we can restore the free memory pointer to before we made
-        // allocations to effectively free the memory. This is safe as the
-        // memory used can be discarded, and the memory pointed to by the free
-        // memory pointer **does not have to point to zero-ed out memory**.
-        // <https://solidity.readthedocs.io/en/v0.6.12/internals/layout_in_memory.html>
-        uint256 freeMemoryPointer;
-        // solhint-disable-next-line no-inline-assembly
-        assembly {
-            freeMemoryPointer := mload(0x40)
-        }
+            // v = uint8(encodedTrade[139])
+            let v := shr(248, calldataload(add(encodedTrade.offset, 139)))
 
-        bytes32 signingDigest;
-        if (v & 0x80 == 0) {
-            // NOTE: The most significant bit **is not set**, so the order is
-            // signed using the EIP-712 sheme, the signing hash is of:
-            // `"\x19\x01" || domainSeparator || orderDigest`.
-            signingDigest = keccak256(
-                abi.encodePacked("\x19\x01", domainSeparator, orderDigest)
-            );
-        } else {
-            // NOTE: The most significant bit **is set**, so the order is signed
-            // using generic message scheme, the signing hash is of:
-            // `"\x19Ethereum Signed Message:\n" || length || data` where the
-            // length is a constant 64 bytes and the data is defined as:
-            // `domainSeparator || orderDigest`.
-            signingDigest = keccak256(
-                abi.encodePacked(
-                    "\x19Ethereum Signed Message:\n64",
-                    domainSeparator,
-                    orderDigest
+            switch and(v, 0x80)
+                case 0 {
+                    mstore(freeMemoryPointer, "\x19\x01")
+                    mstore(add(freeMemoryPointer, 2), domainSeparator)
+                    mstore(add(freeMemoryPointer, 34), orderDigest)
+                    mstore(freeMemoryPointer, keccak256(freeMemoryPointer, 66))
+                }
+                default {
+                    mstore(
+                        freeMemoryPointer,
+                        "\x19Ethereum Signed Message:\n64"
+                    )
+                    mstore(add(freeMemoryPointer, 28), domainSeparator)
+                    mstore(add(freeMemoryPointer, 60), orderDigest)
+                    mstore(freeMemoryPointer, keccak256(freeMemoryPointer, 92))
+                }
+            mstore(add(freeMemoryPointer, 32), and(v, 0x1f))
+            mstore(
+                add(freeMemoryPointer, 64),
+                calldataload(add(encodedTrade.offset, 140))
+            )
+            mstore(
+                add(freeMemoryPointer, 96),
+                calldataload(add(encodedTrade.offset, 172))
+            )
+            // trade.owner = ecrecover(signingDigest, v, r, s)
+            if iszero(
+                staticcall(
+                    gas(),
```

Comment:
Weird that this is needed, since it's not for the version without assembly. I guess this is why you needed to make this function `view`?

Reply:
The `view` is needed for `staticcall`. Unfortunately, there is no way to distinguish a `pure` `staticcall` from a `view staticcall`. That being said, there is no difference between a compiled `view` and `pure` function, except (AFAICT) potential optimizations (that we can't really take advantage of). For example a `pure` function called with a constant parameter can technically be optimized into a constant value. Here we can't do that since we read from `calldata` anyway.